index_release.go

```diff
@@ -19,6 +20,152 @@ type IndexRelease struct {
 	Version     string      `yaml:"version"`
 }
 
+// CompileReleases takes indexReleases and collected version bundles and
+// compiles canonicalized Releases from them.
+func CompileReleases(logger micrologger.Logger, indexReleases []IndexRelease, bundles []Bundle) ([]Release, error) {
+	releases := buildReleases(logger, indexReleases, bundles)
+
+	releases = deduplicateReleaseChangelog(releases)
+
+	return releases, nil
+}
+
+func buildReleases(logger micrologger.Logger, indexReleases []IndexRelease, bundles []Bundle) []Release {
+	bundleCache := make(map[string]Bundle)
+
+	// Create cache of bundles for quick lookup
+	for _, b := range bundles {
+		bundleCache[b.ID()] = b
+	}
+
+	var releases []Release
+
+	for _, ir := range indexReleases {
+		bundles, err := groupBundlesForIndexRelease(ir, bundleCache)
+		if IsBundleNotFound(err) {
+			logger.Log("level", "warning", "message", err.Error())
+			continue
+		}
+
+		if err != nil {
+			panic(err)
```

Comment:
Lets please not panic. We should rather handle the error and return it to the caller so he can decide what to do with it. Just as a general rule. Paniccing at runtime should be the exception. We can do better here.

Reply:
In general I'd have liked to panic if release compilation fails with available data to any other error than missing bundle, but now that you mentioned it, this is actually very wrong thing to do in library. Changed this. Thanks.